File: backend/app/services/traffic.py
"""
Traffic-aware travel duration for Surat.

Resolution order (server-side, no key ever sent to the browser):
  1. TomTom Routing API — real, live/predictive traffic-aware ETA, used when
     TOMTOM_API_KEY is set in the environment (loaded from backend/.env).
  2. Calibrated "Surat Urban Mobility" model — a free offline fallback that
     shapes OSRM's free-flow time with an hour-by-hour congestion curve, a
     speed-aware calibration, and Tapi-bridge queueing. Used when no key is
     configured or the TomTom call fails for any reason.
"""
import os
from datetime import datetime, timedelta

import httpx


def crosses_tapi_river(source: dict, dest: dict) -> bool:
    """True if the trip crosses the Tapi River (lat ~21.200), i.e. North<->South Surat."""
    river_lat = 21.200
    return (source["lat"] > river_lat) != (dest["lat"] > river_lat)


# Per-hour congestion curve for Surat (multiplier on OSRM free-flow time).
# 1.0 = free flow; <1.0 = clearer-than-daytime late-night roads.
HOURLY_CONGESTION = [
    0.90, 0.88, 0.88, 0.88, 0.90, 0.95,   # 00-05
    1.03, 1.18, 1.40, 1.50, 1.34, 1.16,   # 06-11
    1.12, 1.14, 1.10, 1.14, 1.26, 1.46,   # 12-17
    1.60, 1.54, 1.30, 1.12, 1.00, 0.94,   # 18-23
]

# Representative departure hour for each explicit window (used when not 'now').
WINDOW_HOUR = {
    "morning-rush": 9,
    "midday-offpeak": 14,
    "evening-rush": 18,
    "night-freeflow": 1,
    "standard": 11,
}


def simulate_surat_traffic(source: dict, dest: dict, departure_time: str,
                           base_duration_mins: float, distance_km: float = None) -> dict:
    """Calibrated geographic + temporal congestion model (offline fallback)."""
    hour = datetime.now().hour if departure_time == "now" else WINDOW_HOUR.get(departure_time, 11)
    congestion = HOURLY_CONGESTION[hour] if 0 <= hour < 24 else 1.0

    # Speed-aware calibration from OSRM's implied free-flow speed.
    if distance_km and base_duration_mins > 0:
        free_flow_kmh = distance_km / (base_duration_mins / 60)
        if free_flow_kmh < 18:
            congestion = 1 + (congestion - 1) * 0.75   # already slow → dampen extra delay
        elif free_flow_kmh > 35:
            congestion = 1 + (congestion - 1) * 1.15   # open roads → peaks bite harder

    # Tapi River bridge bottleneck, scaled by how congested the hour already is.
    bottleneck_delay = 0
    if crosses_tapi_river(source, dest) and congestion > 1.05:
        bottleneck_delay = round((congestion - 1) * 18)  # ~0 off-peak up to ~11 min at peak

    standard_duration = max(1, round(base_duration_mins))
    traffic_duration = max(2, round(base_duration_mins * congestion + bottleneck_delay))
    delay_mins = max(0, traffic_duration - standard_duration)

    ratio = traffic_duration / standard_duration if standard_duration else 1
    status, color = "🟢 Free Flow", "traffic-light"
    description = "Typical driving conditions with standard road speeds."
    if ratio >= 1.4 or delay_mins >= 10:
        status, color = "🔴 Heavy Congestion", "traffic-heavy"
        description = "Peak commuter congestion. Slow-moving traffic on inner ring roads and arterials."
    elif ratio >= 1.12 or delay_mins >= 3:
        status, color = "🟡 Moderate Traffic", "traffic-moderate"
        description = "Moderate flow with minor delays at major intersections."
    elif congestion < 1:
        description = "Late-night free-flowing speeds. Roads clearer than the daytime baseline."
    if bottleneck_delay >= 3:
        description += f" Includes ~{bottleneck_delay} min Tapi River bridge queue (North<->South Surat)."

    return {
        "isSimulated": True,
        "source": "model",
        "standardDuration": standard_duration,
        "trafficDuration": traffic_duration,
        "delayMins": delay_mins,
        "status": status,
        "color": color,
        "description": description,
    }


def _tomtom_depart_at(departure_time: str):
    """ISO-8601 departAt for explicit windows; None for 'now' (uses live traffic)."""
    if departure_time == "now" or departure_time not in WINDOW_HOUR:
        return None
    now = datetime.now().astimezone()
    hour = WINDOW_HOUR[departure_time]
    target = now.replace(hour=hour, minute=0, second=0, microsecond=0)
    if target < now:
        target = target + timedelta(days=1)
    return target.isoformat()


def fetch_tomtom_traffic(source: dict, dest: dict, departure_time: str,
                         base_duration_mins: float, distance_km: float = None) -> dict:
    """Query the TomTom Routing API for a real traffic-aware ETA."""
    api_key = os.environ.get("TOMTOM_API_KEY", "").strip()
    if not api_key:
        return simulate_surat_traffic(source, dest, departure_time, base_duration_mins, distance_km)

    loc = f"{source['lat']},{source['lng']}:{dest['lat']},{dest['lng']}"
    params = {
        "key": api_key,
        "traffic": "true",
        "travelMode": "car",
        "computeTravelTimeFor": "all",
        "routeType": "fastest",
    }
    depart_at = _tomtom_depart_at(departure_time)
    if depart_at:
        params["departAt"] = depart_at

    url = f"https://api.tomtom.com/routing/1/calculateRoute/{loc}/json"
    try:
        resp = httpx.get(url, params=params, timeout=10.0)
        resp.raise_for_status()
        data = resp.json()
        summary = data["routes"][0]["summary"]

        traffic_secs = summary["travelTimeInSeconds"]
        no_traffic_secs = summary.get("noTrafficTravelTimeInSeconds", traffic_secs)
        delay_secs = summary.get("trafficDelayInSeconds", max(0, traffic_secs - no_traffic_secs))

        traffic_duration = max(1, round(traffic_secs / 60))
        standard_duration = max(1, round(no_traffic_secs / 60))
        delay_mins = max(0, round(delay_secs / 60))
        distance_val = round(summary.get("lengthInMeters", 0) / 1000, 2)

        status, color = "🟢 Free Flow", "traffic-light"
        description = "Live TomTom traffic — roads flowing at normal speeds."
        if delay_mins >= 10:
            status, color = "🔴 Heavy Congestion", "traffic-heavy"
            description = f"Live TomTom traffic — heavy congestion, +{delay_mins} min delay on this route."
        elif delay_mins >= 3:
            status, color = "🟡 Moderate Traffic", "traffic-moderate"
            description = f"Live TomTom traffic — moderate congestion, +{delay_mins} min delay."

        return {
            "isSimulated": False,
            "source": "tomtom",
            "standardDuration": standard_duration,
            "trafficDuration": traffic_duration,
            "delayMins": delay_mins,
            "status": status,
            "color": color,
            "description": description,
            "providerDistance": distance_val,
        }
    except Exception as err:  # noqa: BLE001 — any failure falls back to the offline model
        result = simulate_surat_traffic(source, dest, departure_time, base_duration_mins, distance_km)
        result["fallbackWarning"] = f"TomTom unavailable, using calibrated model (reason: {err})"
        return result


def fetch_traffic(source: dict, dest: dict, departure_time: str,
                  base_duration_mins: float, distance_km: float = None) -> dict:
    """Public entry point: TomTom when a key is configured, else the Surat model."""
    if not source or not dest:
        return None
    return fetch_tomtom_traffic(source, dest, departure_time, base_duration_mins, distance_km)

File: api/traffic.js
/**
 * Vercel Serverless Function: /api/traffic
 *
 * Traffic-aware ETA for Surat, ported from the FastAPI backend so it deploys
 * ON Vercel alongside the frontend — no separate host or tunnel needed.
 *
 *   1. TomTom Routing API — real live/predictive traffic-aware ETA, used when
 *      TOMTOM_API_KEY is set in the Vercel project's environment variables.
 *      The key stays server-side; the browser never sees it.
 *   2. Calibrated "Surat Urban Mobility" model — offline fallback that shapes
 *      OSRM free-flow time with an hourly congestion curve + Tapi-bridge queueing.
 *
 * Response shape matches the old backend exactly, so the frontend is unchanged.
 */

// Surat is IST (UTC+5:30); Vercel runs in UTC, so derive the local hour explicitly.
const IST_OFFSET_MS = 5.5 * 60 * 60 * 1000;

// Per-hour congestion curve for Surat (multiplier on OSRM free-flow time).
const HOURLY_CONGESTION = [
  0.90, 0.88, 0.88, 0.88, 0.90, 0.95,   // 00-05
  1.03, 1.18, 1.40, 1.50, 1.34, 1.16,   // 06-11
  1.12, 1.14, 1.10, 1.14, 1.26, 1.46,   // 12-17
  1.60, 1.54, 1.30, 1.12, 1.00, 0.94    // 18-23
];

const WINDOW_HOUR = {
  'morning-rush': 9,
  'midday-offpeak': 14,
  'evening-rush': 18,
  'night-freeflow': 1,
  'standard': 11
};

const crossesTapiRiver = (source, dest) => {
  const riverLat = 21.200;
  return (source.lat > riverLat) !== (dest.lat > riverLat);
};

/** Calibrated geographic + temporal congestion model (offline fallback). */
const simulateSuratTraffic = (source, dest, departureTime, baseDurationMins, distanceKm = null) => {
  const hour = departureTime === 'now'
    ? new Date(Date.now() + IST_OFFSET_MS).getUTCHours()
    : (WINDOW_HOUR[departureTime] ?? 11);
  let congestion = (hour >= 0 && hour < 24) ? HOURLY_CONGESTION[hour] : 1.0;

  // Speed-aware calibration from OSRM's implied free-flow speed.
  if (distanceKm && baseDurationMins > 0) {
    const freeFlowKmh = distanceKm / (baseDurationMins / 60);
    if (freeFlowKmh < 18) congestion = 1 + (congestion - 1) * 0.75;       // already slow → dampen
    else if (freeFlowKmh > 35) congestion = 1 + (congestion - 1) * 1.15;  // open roads → peaks bite harder
  }

  // Tapi River bridge bottleneck, scaled by how congested the hour already is.
  let bottleneckDelay = 0;
  if (crossesTapiRiver(source, dest) && congestion > 1.05) {
    bottleneckDelay = Math.round((congestion - 1) * 18);
  }

  const standardDuration = Math.max(1, Math.round(baseDurationMins));
  const trafficDuration = Math.max(2, Math.round(baseDurationMins * congestion + bottleneckDelay));
  const delayMins = Math.max(0, trafficDuration - standardDuration);

  const ratio = standardDuration ? trafficDuration / standardDuration : 1;
  let status = '🟢 Free Flow', color = 'traffic-light';
  let description = 'Typical driving conditions with standard road speeds.';
  if (ratio >= 1.4 || delayMins >= 10) {
    status = '🔴 Heavy Congestion'; color = 'traffic-heavy';
    description = 'Peak commuter congestion. Slow-moving traffic on inner ring roads and arterials.';
  } else if (ratio >= 1.12 || delayMins >= 3) {
    status = '🟡 Moderate Traffic'; color = 'traffic-moderate';
    description = 'Moderate flow with minor delays at major intersections.';
  } else if (congestion < 1) {
    description = 'Late-night free-flowing speeds. Roads clearer than the daytime baseline.';
  }
  if (bottleneckDelay >= 3) {
    description += ` Includes ~${bottleneckDelay} min Tapi River bridge queue (North↔South Surat).`;
  }

  return {
    isSimulated: true,
    source: 'model',
    standardDuration,
    trafficDuration,
    delayMins,
    status,
    color,
    description
  };
};

/** ISO-8601 departAt for explicit windows (in IST); null for 'now' (live traffic). */
const computeDepartAt = (departureTime) => {
  if (departureTime === 'now' || !(departureTime in WINDOW_HOUR)) return null;
  const nowMs = Date.now();
  const istNow = new Date(nowMs + IST_OFFSET_MS);
  const hour = WINDOW_HOUR[departureTime];
  // Build the target IST wall-clock instant, then convert back to a true UTC instant.
  let targetMs = Date.UTC(istNow.getUTCFullYear(), istNow.getUTCMonth(), istNow.getUTCDate(), hour, 0, 0) - IST_OFFSET_MS;
  if (targetMs < nowMs) targetMs += 24 * 60 * 60 * 1000;
  return new Date(targetMs).toISOString();
};

/** Query TomTom for a real traffic-aware ETA, falling back to the model on any failure. */
const fetchTomTomTraffic = async (source, dest, departureTime, baseDurationMins, distanceKm, waypoints) => {
  const apiKey = (process.env.TOMTOM_API_KEY || '').trim();
  if (!apiKey) {
    return simulateSuratTraffic(source, dest, departureTime, baseDurationMins, distanceKm);
  }

  // "src:wp1:...:dest" — TomTom routes through the points in order, so a specific
  // corridor gets its own traffic-aware ETA instead of TomTom's single fastest path.
  const points = [source, ...(waypoints || []), dest];
  const loc = points.map((p) => `${p.lat},${p.lng}`).join(':');
  const params = new URLSearchParams({
    key: apiKey,
    traffic: 'true',
    travelMode: 'car',
    computeTravelTimeFor: 'all',
    routeType: 'fastest'
  });
  const departAt = computeDepartAt(departureTime);
  if (departAt) params.set('departAt', departAt);

  const url = `https://api.tomtom.com/routing/1/calculateRoute/${encodeURIComponent(loc)}/json?${params.toString()}`;

  try {
    const resp = await fetch(url);
    if (!resp.ok) throw new Error(`TomTom status ${resp.status}`);
    const data = await resp.json();
    const summary = data.routes[0].summary;

    const trafficSecs = summary.travelTimeInSeconds;
    const noTrafficSecs = summary.noTrafficTravelTimeInSeconds ?? trafficSecs;
    const delaySecs = summary.trafficDelayInSeconds ?? Math.max(0, trafficSecs - noTrafficSecs);

    const trafficDuration = Math.max(1, Math.round(trafficSecs / 60));
    const standardDuration = Math.max(1, Math.round(noTrafficSecs / 60));
    const delayMins = Math.max(0, Math.round(delaySecs / 60));
    const distanceVal = Math.round((summary.lengthInMeters || 0) / 1000 * 100) / 100;

    let status = '🟢 Free Flow', color = 'traffic-light';
    let description = 'Live TomTom traffic — roads flowing at normal speeds.';
    if (delayMins >= 10) {
      status = '🔴 Heavy Congestion'; color = 'traffic-heavy';
      description = `Live TomTom traffic — heavy congestion, +${delayMins} min delay on this route.`;
    } else if (delayMins >= 3) {
      status = '🟡 Moderate Traffic'; color = 'traffic-moderate';
      description = `Live TomTom traffic — moderate congestion, +${delayMins} min delay.`;
    }

    // Run the calibrated model too, so the frontend can show the MEASURED error
    // of our own engine against this live ground-truth ETA.
    const model = simulateSuratTraffic(source, dest, departureTime, baseDurationMins, distanceKm);

    return {
      isSimulated: false,
      source: 'tomtom',
      standardDuration,
      trafficDuration,
      delayMins,
      status,
      color,
      description,
      providerDistance: distanceVal,
      modelTrafficDuration: model.trafficDuration,
      modelStandardDuration: model.standardDuration
    };
  } catch (err) {
    const result = simulateSuratTraffic(source, dest, departureTime, baseDurationMins, distanceKm);
    result.fallbackWarning = `TomTom unavailable, using calibrated model (reason: ${err.message})`;
    return result;
  }
};

export default async function handler(req, res) {
  if (req.method !== 'POST') {
    res.status(405).json({ error: 'Method not allowed' });
    return;
  }

  // Vercel parses JSON bodies, but accept a raw string defensively.
  let body = req.body;
  if (typeof body === 'string') {
    try { body = JSON.parse(body); } catch { body = {}; }
  }
  body = body || {};

  const { source, destination, departureTime = 'now', baseDurationMins, distanceKm = null, waypoints = null } = body;
  if (!source || !destination) {
    res.status(400).json({ error: 'source and destination are required' });
    return;
  }

  try {
    const result = await fetchTomTomTraffic(
      source, destination, departureTime,
      Math.max(1, Number(baseDurationMins) || 1),
      distanceKm != null ? Number(distanceKm) : null,
      Array.isArray(waypoints) ? waypoints : null
    );
    res.status(200).json(result);
  } catch (err) {
    res.status(200).json(simulateSuratTraffic(source, destination, departureTime, Math.max(1, Number(baseDurationMins) || 1), distanceKm));
  }
}
